torchao/prototype/mx_formats/nvfp4_tensor.py

```diff
@@ -0,0 +1,619 @@
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+
+    if not isinstance(weight_tensor, NVFP4Tensor):
+        raise NotImplementedError("NVFP4Tensor: weight must be NVFP4Tensor")
+
+    config = weight_tensor.mm_config
+
+    if config == NVFP4MMConfig.WEIGHT_ONLY:
+        weight_dequant = weight_tensor.to_dtype(weight_tensor._orig_dtype)
+        return torch.nn.functional.linear(input_tensor, weight_dequant, bias)
+    else:
+        input_quant = NVFP4Tensor.to_nvfp4(input_tensor, mm_config=config)
+        return _addmm_nvfp4_dispatch(input_quant, weight_tensor, func, bias=bias)
+
+
+@implements([aten.mm.default, aten.matmul.default])
+def nvfp4_mm(func, types, args, kwargs):
+    input_tensor, weight_tensor = args[0], args[1]
+
+    if not isinstance(weight_tensor, NVFP4Tensor):
+        raise NotImplementedError("NVFP4Tensor: weight must be NVFP4Tensor")
+
+    config = weight_tensor.mm_config
+
+    if config == NVFP4MMConfig.WEIGHT_ONLY:
+        weight_dequant = weight_tensor.to_dtype(weight_tensor._orig_dtype)
+        if isinstance(input_tensor, NVFP4Tensor):
+            input_dequant = input_tensor.to_dtype(input_tensor._orig_dtype)
+            return func(input_dequant, weight_dequant)
+        else:
+            return func(input_tensor, weight_dequant)
+    else:
+        if not isinstance(input_tensor, NVFP4Tensor):
+            input_tensor = NVFP4Tensor.to_nvfp4(input_tensor, mm_config=config)
+        return _addmm_nvfp4_dispatch(input_tensor, weight_tensor, func)
+
+
+@implements([aten.addmm.default])
+def nvfp4_addmm(func, types, args, kwargs):
+    bias, input_tensor, weight_tensor = args[0], args[1], args[2]
+
+    if not isinstance(weight_tensor, NVFP4Tensor):
+        raise NotImplementedError("NVFP4Tensor: weight must be NVFP4Tensor")
+
+    config = weight_tensor.mm_config
+
+    if config == NVFP4MMConfig.WEIGHT_ONLY:
+        weight_dequant = weight_tensor.to_dtype(weight_tensor._orig_dtype)
+        if isinstance(input_tensor, NVFP4Tensor):
+            input_dequant = input_tensor.to_dtype(input_tensor._orig_dtype)
+            return torch.addmm(bias, input_dequant, weight_dequant)
+        else:
+            return torch.addmm(bias, input_tensor, weight_dequant)
+    else:
+        if not isinstance(input_tensor, NVFP4Tensor):
+            input_tensor = NVFP4Tensor.to_nvfp4(input_tensor, mm_config=config)
+        return _addmm_nvfp4_dispatch(input_tensor, weight_tensor, func, bias=bias)
+
+
+def per_tensor_amax_to_scale(amax: torch.Tensor) -> torch.Tensor:
+    """Convert per-tensor amax to per-tensor scale.
+    Used to scale fp32 scales down to fp8 scales
+
+    Args:
+        amax: Per-tensor amax tensor
+
+    Returns:
+        torch.Tensor: Per-tensor scale tensor
+    """
+    return (
+        torch.clamp(amax / F8E4M3_MAX, min=E4M3_EPS, max=F8E4M3_MAX)
+        .to(torch.float8_e4m3fn)
```

Comment:
why is this cast to float8 necessary?

Reply:
Ahhh good catch, this one shouldn't be needed since we only ever use it in fp32